Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift

```diff
@@ -652,6 +652,190 @@ public class SDKSynchronizer: Synchronizer {
+                }
+                
+                // rule out if mismatch of networks
+                guard (info.chainName == "main" && network == .mainnet) 
+                    || (info.chainName == "test" && network == .testnet) else {
+                    continue
+                }
+                
+                // rule out mismatch of consensus branch IDs
+                guard let localBranchID = await blockProcessor.consensusBranchIdFor(Int32(info.blockHeight)) else {
+                    continue
+                }
+
+                guard let remoteBranchID = ConsensusBranchID.fromString(info.consensusBranchID) else {
+                    continue
+                }
+                
+                guard remoteBranchID == localBranchID else {
+                    continue
+                }
+
+                // Rule out servers that are syncing, stuck, or probably on the wrong fork.
+                // To avoid falsely ruling out all servers this can only be a very loose check
+                // (i.e. `ZcashSDK.syncedThresholdBlocks` should not be too small),
+                // because `info.estimatedHeight` may be quite inaccurate.
+                guard info.blockHeight + ZcashSDK.syncedThresholdBlocks >= info.estimatedHeight else {
+                    continue
+                }
+                
+                tmpResults[result.id] = result
+            }
+
+            // rule out all means above latencyThreshold
+            let underThreshold = tmpResults.compactMap {
+                $0.value.mean < (latencyThresholdMillis / 1000.0) ? $0 : nil
+            }
+
+            // sort the server responses by mean
+            let sortedUnderThreshold = underThreshold.sorted {
+                $0.value.mean < $1.value.mean
+            }
+
+            // retain only k servers
+            let sortedUnderThresholdKOnly = sortedUnderThreshold.prefix(3)
+
+            sortedUnderThresholdKOnly.forEach {
+                checkResults[$0.key] = $0.value
+            }
+        }
+
+        // Sequential part
+        var blockResults: [String: CheckResult] = [:]
+
+        for serviceDict in checkResults {
+            guard let info = serviceDict.value.info else {
+                continue
+            }
+            
+            let service = serviceDict.value.service
+            
+            guard info.blockHeight >= nBlocksToFetch else {
+                continue
+            }
+            
+            let stream = service.service.blockStream(startHeight: BlockHeight(info.blockHeight - nBlocksToFetch), endHeight: BlockHeight(info.blockHeight))
+            
+            do {
+                let startTime = Date().timeIntervalSince1970
+                for try await block in stream {
+                    if Date().timeIntervalSince1970 - startTime > fetchThresholdSeconds {
+                        break
+                    }
+                }
+
+                let endTime = Date().timeIntervalSince1970
```

Comment:
Note that there was a [bug report about iterator cleanup](https://forums.swift.org/t/asyncsequence-break-and-cancellation/52787/21) when `break` is used within `for [try] await`. The actual bug ticket, "FB9700937 (break in async for loop does not cancel an async sequence)", appears to have been submitted to the Apple feedback system which is not public (bad :green_apple:!), so I don't know whether it was a real bug or whether it was fixed :frowning_face:.

In any case, if `break` within `for try await` only breaks successfully from the sequence, then the code relies on `Date().timeIntervalSince1970` being monotonic so that a timeout within a loop is guaranteed to imply that `blockTime < fetchThresholdSeconds` is false. In order to not rely on that, it might be better to do something like:

```suggestion
                var endTime = startTime;
                for try await block in stream {
                    endTime = Date().timeIntervalSince1970
                    if endTime - startTime >= fetchThresholdSeconds {
                        break
                    }
                }

                let blockTime = endTime - startTime
```

Reply:
Updated